cpp/src/hash/spark_murmur_hash.cu

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (c) 2019-2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/hashing.hpp>
+#include <cudf/detail/utilities/hash_functions.cuh>
+#include <cudf/detail/utilities/vector_factories.hpp>
+#include <cudf/table/experimental/row_operators.cuh>
+#include <cudf/table/table_device_view.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/tabulate.h>
+
+namespace cudf {
+namespace detail {
+
+namespace {
+
+/**
+ * @brief Computes the hash value of a row in the given table.
+ *
+ * This functor uses Spark conventions for MurmurHash3_32 hashing, which
+ * differs from the standard MurmurHash3_32 implementation. These differences
+ * include:
+ * - specialized tail processing in the hash functor
+ * - specializations for decimal types
+ * - special behavior for serially using the output hash as an input seed for
+ *   the next item
+ * - ignorance of null values: [1], [1, null], and [null, 1] have the same hash
+ *
```

Comment:
Correct me if I'm wrong, but only the third of these four items is actually handled by the adapter, correct? Items 1, 2, and 4 are already handled by the `SparkMurmurHash3_32` hash function? If that's correct, this documentation should more accurately reflect the purpose of just this adapter (handling nested data appropriately).

Reply:
Sort of. The fourth item (ignorance of null values) is implicitly done here by passing the same value for the `seed` and `null_hash` value. That is, it's this class's responsibility to ensure that the seed is returned if the element is null. I can delete the first two.

Reply:
Ah right OK. Yes, please just document the parts that are actually handled here. If you think it's helpful, you can point to the `SparkMurmurHash3_32` for the others. I would also explicitly document that this adapter is only for murmur hashing.